crates/wasmlink/runtime/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+    slab: Slab<Resource>,
+}
+
+#[derive(Debug)]
+struct Resource {
+    wasm: u32,
+    references: u32,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct ResourceIndex(u32);
+
+impl ResourceSlab {
+    pub fn insert(&mut self, wasm: u32) -> ResourceIndex {
+        ResourceIndex(self.slab.insert(Resource {
+            wasm,
+            references: 1,
+        }))
+    }
+
+    pub fn get(&self, idx: ResourceIndex) -> u32 {
+        self.slab.get(idx.0).unwrap().wasm
+    }
+
+    pub fn clone(&mut self, idx: ResourceIndex) {
+        let resource = self.slab.get_mut(idx.0).unwrap();
+        resource.references = resource.references.checked_add(1).unwrap();
+    }
+
+    pub fn drop(&mut self, idx: ResourceIndex) -> Option<u32> {
+        let resource = self.slab.get_mut(idx.0).unwrap();
+        assert!(resource.references > 0);
+        resource.references -= 1;
+        if resource.references != 0 {
+            return None;
+        }
+        let resource = self.slab.remove(idx.0).unwrap();
+        Some(resource.wasm)
+    }
+}
+
+lazy_static::lazy_static! {
+    static ref SLABS: Mutex<Vec<(IndexSlab, ResourceSlab)>> = Mutex::new(Vec::new());
+}
+
+#[no_mangle]
+pub extern "C" fn resource_insert(id: u32, res: u32) -> u32 {
+    let mut slabs = SLABS.lock().unwrap();
+
+    if slabs.len() <= id as usize {
+        slabs.resize_with(id as usize + 1, Default::default);
+    }
+
+    let (indexes, resources) = slabs.get_mut(id as usize).unwrap();
+    let index = resources.insert(res);
+    indexes.insert(index)
+}
+
+#[no_mangle]
+pub extern "C" fn resource_get(id: u32, idx: u32) -> u32 {
+    let mut slabs = SLABS.lock().unwrap();
+
+    if slabs.len() <= id as usize {
+        slabs.resize_with(id as usize + 1, Default::default);
+    }
+
+    let (indexes, resources) = slabs.get(id as usize).unwrap();
+    let res_idx = indexes.get(idx).unwrap();
+    resources.get(res_idx)
+}
+
+#[no_mangle]
+pub extern "C" fn resource_clone(id: u32, idx: u32) -> u32 {
+    let mut slabs = SLABS.lock().unwrap();
+
+    if slabs.len() <= id as usize {
+        slabs.resize_with(id as usize + 1, Default::default);
+    }
+
+    let (indexes, resources) = slabs.get_mut(id as usize).unwrap();
+    let res_idx = indexes.get(idx).unwrap();
+    resources.clone(res_idx);
+    indexes.insert(res_idx)
+}
+
+#[no_mangle]
+pub extern "C" fn resource_remove(id: u32, idx: u32) -> u64 {
+    let mut slabs = SLABS.lock().unwrap();
+
+    if slabs.len() <= id as usize {
+        slabs.resize_with(id as usize + 1, Default::default);
+    }
+
+    let (indexes, resources) = slabs.get_mut(id as usize).unwrap();
+    let res_idx = indexes.remove(idx).unwrap();
+
+    // The return value's upper 32-bits is a flag to denote if the resource is still alive.
+    // If the upper 32-bits are 0, the lower 32-bits are expected to be the resource to drop.
+    match resources.drop(res_idx) {
+        Some(wasm) => wasm as u64,
+        None => 1u64 << 32,
```

Comment:
An alternative design is that instead of having callers check this flag the initialization of this module appends closures to its function table, and the function pointer is passed in as an argument here. That way this would perform, when the resource is actually dropped, a dynamic dispatch to the destructor.

Not that it's really any better one way or another, just an alternate possibility!

Reply:
I hadn't considered a `funcref` approach; I'll see what I can make happen.

Reply:
Oh sorry to be clear I wouldn't use `funcref` since Rust has no support for that. Rather the functions are inserted into the funcref table exported by this module (via the adapter glue) and then the index of the function to call is passed here (which is the representation of a function pointer in Rust anyway).

This would require modifying the source wasm, though, since IIRC LLD fills in the min/max of the funcref table as the same value, so you'd have to increase the max at least to be able to insert functions into it. I don't think any of this is really worth it to do at this time when what you've got here is already working.

Reply:
Ah now I understand; yeah, given the complexity of that approach (i.e. having to post-process the runtime module), perhaps keeping a simple "branch if the upper 32-bits is non-zero" approach I have here will suffice for now.